### DISCUSSION TITLE
Help with actions for Ruby gem

### DISCUSSION THREAD

**Discussion by janosrusiczki** [community/community#25851] [Actions]
<p>I have a Ruby gem which I’ve been building with Travis but I want to make the transition to GitHub actions. I took the example Ruby gem action and expanded upon it somewhat, but I feel it could be further improved. The file as it stands now:</p>
<p><a href="https://github.com/haiafara/underpass/blob/04f2c48a20dc66af051a6c9e5ccd91465387eaac/.github/workflows/gempush.yml" rel="nofollow noopener">https://github.com/haiafara/underpass/blob/04f2c48a20dc66af051a6c9e5ccd91465387eaac/.github/workflows/gempush.yml</a></p>
<p>As you can see I have three jobs:</p>
<ul>
<li>rubocop - checks for code style</li>
<li>rspec - depends on rubocop and runs the test suite</li>
<li>publish - depends on rspec and publishes the gem if the action was run on the master branch</li>
</ul>
<p>What I would like to improve:</p>
<ul>
<li>All three jobs run the checkout and Ruby setup steps, is it possible to run these only one time? I mean once rubocop finishes it would preserve the project directory and the next step would continue from that point.</li>
<li>Is it possible to somehow separate the publish job so that it doesn’t even start on branches that are different from master?</li>
</ul>

**Comment by wipe2238** [comment, marked as answer]
<blockquote>
<p><a class="mention" href="/janosrusiczki">@janosrusiczki</a> wrote:<br>
preserve the project directory and the next step would continue from that point.</p>
</blockquote>
<p>No caching available yet…</p>
<p>EDIT: Posted too fast… next step or next job?</p>
<p>Everything is in place until job finishes, so everything* you do in one step is visible in another.</p>
<p>If it’s about next job, currently only way to pass files from job to job is to upload files in one job and download in another (for example using actions/upload-artifact + actions/download-artifact).</p>
<p>*Current directory changes on each step; if you <code>cd my_dir</code> in one step, next one will push you back where you started.</p>
<blockquote>
<p><a class="mention" href="/janosrusiczki">@janosrusiczki</a> wrote:</p>
</blockquote>
<blockquote>
<ul>
<li>Is it possible to somehow separate the publish job so that it doesn’t even start on branches that are different from master?</li>
</ul>
</blockquote>
<p>You have a small mistake there;</p>
<pre><code>if: github.ref == 'master'
</code></pre>
<p>should be</p>
<pre><code>if: github.ref == 'refs/heads/master'
</code></pre>
<p>If you want to skip whole job, move that check under <code>job.\&lt;id\&gt;</code></p>
<pre><code>publish:
 runs-on: ubuntu-latest
 if: github.ref == 'refs/heads/master'
</code></pre>

**Comment by janosrusiczki** [comment]
<p>Thank you for the hints. I moved the if statement and indeed it’s skipping that job altogether.</p>
<p>As for artifact uploading and downloading: I printed out where it is installing the gems and it said: /opt/hostedtoolcache/Ruby/2.6.3/x64/lib/ruby/gems/2.6.0/gems - so I thought I’ll upload that directory as an artifact and download it in the next job. The problem is that that directory is huge, because as its name says it’s basically a systemwide cache. And since the upload was progressing really slowly I canceled it, but I’m not sure it would have worked anyway because AFAIK bundle install does some linking too.</p>
<p>The good news is that as stated <a href="https://github.com/actions/toolkit/issues/47" rel="nofollow noopener">here</a> they’re currently working on caching, so I won’t try reinvent the wheel, I’ll just wait patiently.</p>
